### ISSUE TITLE
ESLint aborts early with ecmaFeatures { "modules": true } 

### ISSUE THREAD

**Issue by feross** [eslint/eslint#3128]
Config:

```json
{
  "ecmaFeatures": {
    "modules": true
  },
  "rules": {
    "space-before-function-paren": [2, "always"]
  }
}
```

Code:

```js
function fun () {}

fun(function(err, _, _, what) {
})
```

In v0.24.1, this works as exepcted and returns `t.js:4:13: Missing space before function parentheses.`.

In v1.0.0-rc1, this aborts early with the message `t.js:4:-10: Strict mode function may not have duplicate parameter names` and an invalid column number of `-10`.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by btmills** [comment]
The error message is correct, though the column number looks like an espree issue: eslint/espree#173.

**Comment by feross** [comment]
In v1.0.0-rc1, this error causes eslint to abort early and report this single error, skipping all other errors. In v0.24.1, this error was not fatal.

**Comment by IanVS** [comment]
I believe it should indeed be fatal, since it is a parsing error and not a linting error.  The 0.24.1 case sounds like a bug.  Or are you saying it's somehow related to `ecmaFeatures { "modules": true }` as the issue title suggests?

**Comment by btmills** [comment]
I just checked v0.24.1 and saw the same behavior as v1.0.0-rc-1. Did you have an older version of espree installed under v0.24.1? Anytime ESLint encounters a parse error, it reports just that error since that means there's no AST for the rules to check.

**Comment by nzakas** [comment]
Fixed Espree error location bug: https://github.com/eslint/espree/releases/tag/v2.2.3

**Comment by feross** [comment]
It may be a parse error in ES2015 terms, but there's no reason that espree couldn't return a parse error and an AST in this case. It's not "fatal" in the sense that espree doesn't have enough information to construct a complete AST.

It's obviously your call, but from a UX perspective, it'd be nice to avoid the scenario where the user thinks they only have one error, fix the error, re-run the linter, and suddenly they have 10 errors.

**Comment by dcousens** [comment]
> It's obviously your call, but from a UX perspective, it'd be nice to avoid the scenario where the user thinks they only have one error, fix the error, re-run the linter, and suddenly they have 10 errors.

I have to agree with this entirely,  1 warning from the linter is easily ignored,  1000's not so much.
This potentially stops 1000's from ever popping up.

**Comment by michaelficarra** [comment]
Espree should support disabling early error checks. See [my writeup on how I did this](http://engineering.shapesecurity.com/2015/04/two-phase-parsing.html) in the Shift parser.

**Comment by nzakas** [comment]
The error is already thrown as a tolerant error, but we don't run in tolerant mode. We could give that a try, I'm just not sure the effort involved.

**Comment by dcousens** [comment]
Is this classifiable as a bug?

**Comment by IanVS** [comment]
@dcousens what would the bug be?  I believe it is operating as intended.

**Comment by feross** [comment]
Bug: Source code has 100 errors, eslint only reports 1 error.

**Comment by dcousens** [comment]
@IanVS,  more succinctly,  `eslint` is marking the outlined parsing case as a show-stopper to all other parsing, when in fact it could continue parsing quite easily.

**Comment by ilyavolodin** [comment]
I opened a new issue #3815 to cover proposed solution. Closing this issue, because this is currently behaving as intended.
